lib/mobility-core/src/Kernel/Utils/SlidingWindowCounters.hs

```diff
@@ -245,26 +245,79 @@ decrementByValueImpl mbTimeStamp val keyModifier key SlidingWindowOptions {..} =
   ( L.MonadFlow m,
     Redis.HedisFlow m r
   ) =>
   Text ->
   (Text -> Text) ->
   (Text -> Text) ->
+  (Text -> Text -> Text) ->
+  Text ->
   SlidingWindowOptions ->
   m Double
-getLatestRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn s@SlidingWindowOptions {..} = do
+calculateRatioForFirstTime driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType s@SlidingWindowOptions {..} = do
   utcTime <- L.runIO getCurrentTime
   let positiveCaseKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkPostiveCaseKeyfn driverId)
-  let totalCountKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkTotalCaseKeyfn driverId)
+      totalCountKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkTotalCaseKeyfn driverId)
   positiveCases <- sum <$> mapMaybeM Redis.get positiveCaseKeysList
   totalCases <- nonZero . sum <$> mapMaybeM Redis.get totalCountKeysList
-  pure $ positiveCases / totalCases
+  let ratio = positiveCases / totalCases
+      expTime = fromInteger . floor $ diffUTCTime (incrementPeriod periodType utcTime) utcTime
+  Redis.setExp (mkOldRatioKeyfn driverId ratioType) ratio expTime
+  pure ratio
   where
     nonZero :: Double -> Double
     nonZero a = if a == 0.0 then 1.0 else a
 
+recalculateRatio ::
+  ( L.MonadFlow m,
+    Redis.HedisFlow m r
+  ) =>
+  Text ->
+  (Text -> Text) ->
+  (Text -> Text) ->
+  (Text -> Text -> Text) ->
+  Text ->
+  SlidingWindowOptions ->
+  m Double
+recalculateRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType s@SlidingWindowOptions {..} = do
+  let positiveCaseKey = mkPostiveCaseKeyfn driverId
+      totalCaseKey = mkTotalCaseKeyfn driverId
+      oldRatioKey = mkOldRatioKeyfn driverId ratioType
+  positiveCase <-
+    Redis.get positiveCaseKey >>= \case
+      Nothing -> return 0
+      Just num -> return num
+  totalCase <-
+    Redis.get totalCaseKey >>= \case
+      Nothing -> return 1
+      Just total -> return total
+  ratio <-
+    Redis.get oldRatioKey >>= \case
+      Nothing -> calculateRatioForFirstTime driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType s
+      Just prevRatio -> return prevRatio
+  let newRatio = (positiveCase / totalCase + ratio) / 2
+  return newRatio
+
+getLatestRatio ::
+  ( L.MonadFlow m,
+    Redis.HedisFlow m r
+  ) =>
+  Text ->
+  (Text -> Text) ->
+  (Text -> Text) ->
+  (Text -> Text -> Text) ->
+  Text ->
+  SlidingWindowOptions ->
+  m Double
+getLatestRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType SlidingWindowOptions {..} = do
+  ratio :: (Maybe Double) <- Redis.get (mkOldRatioKeyfn driverId ratioType)
+  case ratio of
+    Nothing -> calculateRatioForFirstTime driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType SlidingWindowOptions {..}
+    _ -> recalculateRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType SlidingWindowOptions {..}
```

Comment:
you can send the cached value to `recalculateRatio` function instead of fetching the value inside again.